Source/DafnyCore/Compilers/SinglePassCompiler.cs

```diff
@@ -1047,9 +1049,134 @@ public class TypeArgumentInstantiation {
     protected abstract ILvalue EmitMemberSelect(Action<ConcreteSyntaxTree> obj, Type objType, MemberDecl member, List<TypeArgumentInstantiation> typeArgs, Dictionary<TypeParameter, Type> typeMap,
       Type expectedType, string/*?*/ additionalCustomParameter = null, bool internalAccess = false);
 
-    protected void EmitArraySelect(string index, Type elmtType, ConcreteSyntaxTree wr) {
-      EmitArraySelect(new List<string>() { index }, elmtType, wr);
+    protected enum MemberCompileStatus { Ordinary, Identity, AlwaysTrue }
+
+    protected MemberCompileStatus GetMemberStatus(MemberDecl member) {
+      if (member.EnclosingClass is DatatypeDecl dt) {
+        if (OptimizesInvisibleDatatypeWrappers && IsInvisibleWrapper(dt, out var dtor) && dtor == member) {
+          // "member" is the sole destructor of an invisible datatype wrapper
+          return MemberCompileStatus.Identity;
+        } else if (member is DatatypeDiscriminator) {
+          // a discriminator of an inductive or coinductive datatype
+          return dt.Ctors.Count(c => !c.IsGhost) == 1 ? MemberCompileStatus.AlwaysTrue : MemberCompileStatus.Ordinary;
+        }
+      }
+
+      return MemberCompileStatus.Ordinary;
     }
+
+    protected bool IsInvisibleWrapper(DatatypeDecl dt, out DatatypeDestructor coreDestructor) {
```

Comment:
What does `invisible` in the name mean? Could we rename this to `ShouldDatatypeBeErased` ?

Reply:
With "invisible wrapper", I meant that the single compiled constructor is "erasable". I considered renaming it `IsEraseableDatatype`, but is a bit misleading, since it's only the "wrapper" part that's erased--not the entire datatype value. I'll rename it `IsEraseableDatatypeWrapper` and will do the same renaming for the test file and command-line option.